### PR TITLE
Unavailable products

### PR DESCRIPTION
**Pull request by mihawagner**

### CODE REVIEW

File: src/Commands/AmazonProductWidgetCommands.php [hunk 1]
```diff
@@ -279,4 +300,42 @@ public function getProductInfo(string $asin, array $options = ['renew' => FALSE]
...
+      'title'        => t('Updating Product Node mapping'),
+      'init_message' => t('Preparing to sync @count nodes...'),
+      'finished'     => [BatchProductMapUpdateService::class, 'finished'],
+    ];
+
+    try {
+      $nodeIds = $this->entityTypeManager->getStorage('node')
+        ->getQuery()
+        ->execute();
+      $totalNodes = count($nodeIds  );
```
Comment (mbm80) [inline src/Commands/AmazonProductWidgetCommands.php]: o_O

File: src/ProductService.php [hunk 2]
```diff
@@ -745,7 +745,15 @@ public function getProductsWithFallback(AmazonProductField $product_field, NodeI
     // Replace unavailable products with ones from the search term fallback.
     $replace = [];
     foreach ($product_data as $asin => $data) {
-      $callbackResults = $this->moduleHandler->invokeAll('amazon_product_widget_alter_validate_product_data', [$product_field, $data]);
+      $callbackResults = [];
+
+      if ($data !== FALSE) {
+        $callbackResults = $this->moduleHandler->invokeAll('amazon_product_widget_alter_validate_product_data', [$product_field, $data]);
+        $this->productStore->setAvailable($asin, TRUE);
```
Comment (mbm80) [inline src/ProductService.php]: hm, this should be further down below where $valid is set, or why is this here?

File: amazon_product_widget.install [hunk 3]
```diff
@@ -122,6 +127,30 @@ function amazon_product_widget_schema() {
     ],
   ];
 
+  $schema['amazon_product_widget_asin_map'] = [
+    'description' => 'Stores ASINs that are used in entities.',
+    'fields' => [
+      'entity_id' => [
```
Comment (mbm80) [inline amazon_product_widget.install:135]: entity_id & entity_type should be an index (non unique)

File: amazon_product_widget.install [hunk 4]
```diff
@@ -122,6 +127,30 @@ function amazon_product_widget_schema() {
...
+        'not null' => TRUE,
+        'unsigned' => TRUE,
+      ],
+      'entity_type' => [
+        'description' => 'Entity type',
+        'type' => 'varchar_ascii',
+        'length' => 64,
+        'not null' => TRUE,
+      ],
+      'asin' => [
```
Comment (mbm80) [inline amazon_product_widget.install:147]: asin should be an index (non unique)

Comment (mbm80) [reply amazon_product_widget.install:147]: and amazon_product_widget_key_value.name should be an index (non unique)

File: src/ProductUsageService.php [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+            continue;
+          }
+
+          $asins = array_merge($asins, $productField->getAsins());
+        }
+        catch (\Exception $exception) {
+          watchdog_exception('amazon_product_widget', $exception);
+        }
+      }
+      elseif ($definition->getType() === 'entity_reference_revisions') {
```
Comment (mbm80) [inline src/ProductUsageService.php]: but could also be an entity_reference (w/o revision)

File: src/ProductUsageService.php [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+        /** @var \Drupal\paragraphs\ParagraphInterface $paragraphField */
+        $paragraphField = $entity->get($fieldName);
```
Comment (mbm80) [inline src/ProductUsageService.php]: could be any referenced entity, doesn't have to be a paragraph

File: src/ProductUsageService.php [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+        if (!$paragraphField) {
+          continue;
+        }
+        $paragraphs = $paragraphField->referencedEntities();
```
Comment (mbm80) [inline src/ProductUsageService.php]: ... so i would just call this $referenced_entity

File: src/ProductUsageService.php [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+      return;
+    }
+
+    // MenuLinkEntity returns NULL on call to $entity->id() so we check here to
+    // be sure.
+    if ($entity->id() === NULL) {
+      return;
+    }
+
+    $this->purge($entity->id(), $entity->getEntityTypeId());
```
Comment (mbm80) [inline src/ProductUsageService.php]: so this will fragment the database table, since this is run frequently we should add a cron job, eg. daily, or maybe weekly to optimize table

Comment (mbm80) [reply src/ProductUsageService.php]: also before purging, first we should check whether an update (purge/insert) is even necessary, ie. any asins have changed in the first place

File: src/ProductUsageService.php [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+              }
+            }
+          }
+        }
+      }
+    }
+
+    $asins = array_merge($asins, $hookAsins);
+    $asins = array_unique($asins);
+    $this->insert($entity->id(), $entity->getEntityTypeId(), $asins);
```
Comment (mbm80) [inline src/ProductUsageService.php]: so purge above this line and check before existing entries with the results here

File: amazon_product_widget.install (hunk 4 above)
Comment (mbm80) [reply amazon_product_widget.install:147]: > and amazon_product_widget_key_value.name should be an index (non unique)

actually, name + available combined index makes more sense

File: README.md [hunk 10]
```diff
@@ -209,6 +210,18 @@ file will be downloaded from the Amazon API (if configured).
...
+
+* `apw:entity-products <entity_id> <entity_type>`
+
+Returns products that are contained in the given entity. `<entity_type>` is optional and defaults to 'node'.
+
+* `apw:product-entities <ASIN>`
+
+Returns the entity ID and type that contain the given product (ASIN).
+
 ## Hooks
```
Comment (mbm80) [inline README.md:225]: missing the new hook

File: amazon_product_widget.api.php [hunk 11]
```diff
@@ -35,3 +36,19 @@ function hook_amazon_product_widget_alter_product_data(array &$products_containe
...
+ * Allows you to alter the ASINs returned for the entity. This could be useful
+ * if they are stored in a different place than the AmazonProductField.
+ *
+ * @param \Drupal\Core\Entity\FieldableEntityInterface $entity
+ *   The entity that is being searched.
+ *
+ * @return array
+ *   Array of ASINs that were found in the entity. Can be empty.
+ */
+function hook_amazon_product_widget_alter_asin_map(FieldableEntityInterface $entity) {
```
Comment (mbm80) [inline amazon_product_widget.api.php]: its not an alter hook, lets call it .._get_custom_asin_map or so

File: amazon_product_widget.install [hunk 12]
```diff
@@ -40,6 +40,12 @@ function amazon_product_widget_schema() {
         'description' => 'Stores overrides for amazon product widget',
         'not null' => FALSE,
       ],
+      'available' => [
+        'description' => 'Represents product availability',
+        'type' => 'int',
+        'size' => 'tiny',
+        'default' => 0,
```
Comment (mbm80) [inline amazon_product_widget.install:47]: if we do default 0, which makes sense, id suggest setting it to 1 for all existing entries in an update hook, and let the cron job slowly work out the unavailable ones.

File: amazon_product_widget.install [hunk 13]
```diff
@@ -68,7 +75,6 @@ function amazon_product_widget_schema() {
         'description' => 'UNIX timestamp when this deal was created.',
         'type' => 'int',
         'not null' => TRUE,
-        'default' => 0,
```
Comment (mbm80) [inline amazon_product_widget.install]: should this be removed?

File: amazon_product_widget.install [hunk 14]
```diff
@@ -51,6 +57,7 @@ function amazon_product_widget_schema() {
     'indexes' => [
       'all' => ['name', 'collection', 'renewal'],
       'renewal' => ['renewal'],
+      'name_available' => ['name', 'available'],
```
Comment (mbm80) [inline amazon_product_widget.install:60]: index for existing tables also must be updated in an update hook

File: amazon_product_widget.install (hunk 13 above)
Comment (mihawagner) [reply amazon_product_widget.install]: o_O gremlins... Thanks!

**Comment by mbm80** [comment]
works like a charm
